### PR TITLE
Kyle/test big file v3

### PR DESCRIPTION
**Pull request by kyleryan207**

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# This file defines code ownership for the repository
+# Each line is a file pattern followed by one or more owners
+
+# Global owner for everything in the repository
+* @kyleryan207
```
Comment (kyleryan207) [inline CODEOWNERS:5]: Missing newline at end of file. This is a common convention and some tools expect it.

File: check.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      // if (owner !== config.GITHUB_OWNER || repo !== config.GITHUB_REPO) {
+      //   logger.info('Ignoring PR from different repository', {
+      //     owner,
+      //     repo,
+      //     configured: `${config.GITHUB_OWNER}/${config.GITHUB_REPO}`,
+      //   });
+      //   return res.status(200).json({ message: 'Repository not configured for review' });
+      // }
+
+      this.processReviewAsync(repository.owner.login, repository.name, number);
```
Comment (kyleryan207) [inline check.ts:45]: This async operation is started but not awaited, which means the response is sent immediately without waiting for the review to complete. This could lead to race conditions or make error handling difficult. Consider either awaiting this operation or clearly documenting this fire-and-forget behavior.

```suggestion
await this.processReviewAsync(repository.owner.login, repository.name, number);
```

File: check.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    } catch (error: any) {
+      logger.error('Error handling PR webhook:', { error: error.message });
```
Comment (kyleryan207) [inline check.ts:47]: Missing return statement after sending the response. Without it, execution continues and could potentially send another response, causing 'Cannot set headers after they are sent' errors.

```suggestion
      this.processReviewAsync(repository.owner.login, repository.name, number);
      return res.status(200).json({ message: 'Review started' });
```

File: check.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      res.status(500).json({ error: 'Internal server error' });
```
Comment (kyleryan207) [inline check.ts:48]: This catch block will only handle synchronous errors in the try block, not errors from the async `processReviewAsync` call since it's not awaited. Errors in the async operation will be logged but won't affect the HTTP response.

File: check.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
+import { Request, Response } from 'express';
```
Comment (kyleryan207) [inline check.ts:1]: Missing test coverage for this webhook controller. Given the critical nature of webhook handling (security, reliability), comprehensive tests should cover: webhook signature validation, different event types, error scenarios, async processing behavior, and edge cases. Consider adding unit tests and integration tests.

File: check.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      });
+
+      // Only process certain actions
+      const reviewActions = ['opened', 'synchronize', 'reopened'];
+      if (!reviewActions.includes(action)) {
+        logger.info('Ignoring PR action', { action });
+        return res.status(200).json({ message: 'Action ignored' });
+      }
+
+      // Only process PRs for configured repository
```
Comment (kyleryan207) [inline check.ts:32]: The commented-out repository validation logic should either be implemented or removed. If this is meant for multi-repo support later, consider adding a TODO comment explaining the future plans. If not needed, clean up the code.

```suggestion
      // TODO: Add repository validation when multi-repo support is needed
      // const owner = repository.owner.login;
      // const repo = repository.name;
```

File: check.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+import config from '@/utils/config';
+import logger from '@/utils/logger';
+
+class WebhookController {
+  private reviewService: ReviewService;
+
+  constructor() {
+    this.reviewService = new ReviewService();
+  }
+
```
Comment (kyleryan207) [inline check.ts:14]: Consider extracting the webhook request type definitions to improve type safety. The current generic typing doesn't provide compile-time validation of the expected webhook payload structure.

**Review by kyleryan207 — COMMENTED** [review]
## 🤖 AI Code Review

**Overall Assessment**: 🚨 SECURITY_CONCERN

**Summary**: This PR introduces a webhook controller for handling GitHub PR review automation, but has significant security vulnerabilities and architectural concerns that need immediate attention. While the code shows good structure with proper TypeScript typing and error handling patterns, critical security measures are missing.

## 📋 Code Quality (6/10)

**Issues**:
- Commented out repository validation allows any repository to trigger webhooks
- Missing webhook signature verification makes the endpoint vulnerable to spoofing
- Direct database calls in controller violate the documented pattern of using service functions
- Fire-and-forget async processing with processReviewAsync() could lead to silent failures
- Missing input validation for webhook payloads
- Error responses may leak internal information

**Suggestions**:
- Follow the documented pattern: 'Do not make DB calls directly in Controller functions - All DB calls should be wrapped in a service function with clear typing'
- Add proper logging and monitoring for the async review process
- Consider using a job queue for background processing instead of fire-and-forget
- Add request timeout handling for webhook processing
- Implement structured error responses that don't expose internal details

## 📝 Acceptance Criteria
**Status**: ❌ Not Met

No specific acceptance criteria provided, but the implementation lacks essential security measures and violates documented architectural patterns. The code appears to be a work-in-progress based on the commented-out validation logic.

## 💬 Detailed Feedback

Welcome to the team! 🎉 You've built a solid foundation for a webhook controller with good TypeScript practices and proper error handling structure. However, there are some critical security and architectural issues we need to address before this can be merged.

**🔒 Critical Security Issues:**

The biggest concern is the missing webhook signature verification. GitHub signs all webhook payloads with a secret, and we must verify this signature to ensure the webhook actually came from GitHub. Without this, anyone could send fake webhooks to your endpoint. Here's how to implement it:

```typescript
import crypto from 'crypto';

function verifyGitHubSignature(payload: string, signature: string, secret: string): boolean {
  const expectedSignature = crypto
    .createHmac('sha256', secret)
    .update(payload)
    .digest('hex');
  return crypto.timingSafeEqual(
    Buffer.from(`sha256=${expectedSignature}`),
    Buffer.from(signature)
  );
}
```

Also, the commented-out repository validation (lines 32-43) should be uncommented and properly configured to prevent unauthorized repositories from triggering reviews.

**🏗️ Architecture Alignment:**

Our codebase follows a specific pattern documented in the style guide: "Do not make DB calls directly in Controller functions - All DB calls should be wrapped in a service function with clear typing." Your `ReviewService` usage is great and follows this pattern correctly!

However, the `processReviewAsync()` method uses fire-and-forget processing, which could lead to silent failures. Consider using a proper job queue or at least ensuring errors are properly tracked and reported.

**📚 Learning Resources:**
- [GitHub Webhook Security Documentation](https://docs.github.com/en/webhooks-and-events/webhooks/securing-your-webhooks)
- [Express.js Security Best Practices](https://expressjs.com/en/advanced/best-practice-security.html)

**🔧 Next Steps:**
1. Add webhook signature verification
2. Uncomment repository validation with proper configuration
3. Add rate limiting middleware
4. Improve error handling for async processing
5. Add input validation for webhook payloads

The code structure and TypeScript usage are excellent - you clearly understand our patterns well. Once these security measures are in place, this will be a solid addition to the codebase!

## ⚡ Action Required

Implement webhook signature verification and uncomment repository validation before this can be safely merged. Address other security recommendations for production readiness.


---
*This review was generated by AI. Please use human judgment for final decisions.*

File: check.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      //   return res.status(200).json({ message: 'Repository not configured for review' });
+      // }
+
+      this.processReviewAsync(repository.owner.login, repository.name, number);
+    } catch (error: any) {
+      logger.error('Error handling PR webhook:', { error: error.message });
+      res.status(500).json({ error: 'Internal server error' });
+    }
+  }
+
```
Comment (kyleryan207) [inline check.ts:51]: This method is doing heavy work but lacks proper error handling propagation. Consider making this method return a result object that indicates success/failure, which could be useful for monitoring or future webhook response handling.

File: check.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  async handleIssueComment(req: Request<{}, {}, IssueCommentWebhookEvent>, res: Response): Promise<void> {
+    try {
+      const { action, comment, issue, repository } = req.body;
+
+      // Only process if it's a PR comment (issues and PRs share the same endpoint)
+      if (!issue.pull_request) {
+        res.status(200).json({ message: 'Not a PR comment' });
+        return;
```
Comment (kyleryan207) [inline check.ts:103]: Early return pattern is good here, but consider adding a log statement to track when non-PR comments are received for monitoring purposes.

```suggestion
      if (!issue.pull_request) {
        logger.debug('Ignoring non-PR comment', { issueNumber: issue.number });
        res.status(200).json({ message: 'Not a PR comment' });
        return;
      }
```

File: check.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      logger.error('Error handling issue comment webhook:', { error: error.message });
+      res.status(500).json({ error: 'Internal server error' });
+    }
+  }
+
+  async healthCheck(_req: Request, res: Response): Promise<void> {
+    try {
+      const health = {
+        status: 'healthy',
+        timestamp: new Date().toISOString(),
```
Comment (kyleryan207) [inline check.ts:126]: Using `process.env.npm_package_version` may not work in production builds where package.json isn't available. Consider setting this via build-time environment variable or config file.

```suggestion
        version: config.VERSION || process.env.npm_package_version || '1.0.0',
```

File: check.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        environment: config.NODE_ENV,
+      };
+
+      res.status(200).json(health);
+    } catch (error: any) {
+      logger.error('Health check error:', { error: error.message });
+      res.status(500).json({ status: 'unhealthy', error: error.message });
+    }
+  }
+}
```
Comment (kyleryan207) [inline check.ts:137]: Missing export of controller instance. Typically you'd export an instance for use in routes:

```suggestion
export default WebhookController;

// Export instance for direct use in routes
export const webhookController = new WebhookController();
```

**Review by kyleryan207 — COMMENTED** [review]
## 🤖 AI Code Review

**Overall Assessment**: ⚠️ NEEDS_CHANGES

**Summary**: This PR introduces webhook handling for GitHub events, but has critical security vulnerabilities and several code quality issues that need addressing before merge. The implementation shows good understanding of async patterns but lacks essential security measures.

## 📋 Code Quality (4/10)

**Issues**:
- Missing webhook signature verification - GitHub webhooks should be validated using HMAC-SHA256
- Async operation started without awaiting in handlePullRequest method (line 44)
- Missing return statement after sending response, risking multiple responses
- Commented-out repository validation logic should be implemented or removed
- No test coverage for critical webhook functionality
- Generic typing instead of proper request type validation
- Error handling only catches synchronous errors, not async ones from processReviewAsync

**Suggestions**:
- Add webhook signature verification using GitHub's secret
- Await the processReviewAsync call or implement proper fire-and-forget pattern
- Add return statement after res.status(200).json() calls
- Implement repository validation or remove commented code
- Add comprehensive test coverage for webhook handlers
- Extract webhook request type definitions for better type safety
- Use proper async/await error handling throughout

## 📝 Acceptance Criteria
**Status**: ❌ Not Met

Without a PR description or Linear requirements, we can't verify specific acceptance criteria. However, basic webhook security and reliability standards are not met.

## 💬 Detailed Feedback

Welcome to the team! This is a solid foundation for webhook handling, but there are some important improvements needed, especially around security.

**Critical Security Issue:**
The most important issue is missing webhook signature verification. GitHub signs all webhook payloads with a secret, and we must verify this signature to ensure the requests are legitimate. Without this, anyone could send fake webhook requests to trigger reviews. Here's how to implement it:

```typescript
import crypto from 'crypto';

function verifySignature(payload: string, signature: string, secret: string): boolean {
  const hmac = crypto.createHmac('sha256', secret);
  const digest = 'sha1=' + hmac.update(payload).digest('hex');
  return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(digest));
}
```

**Async Pattern Issue:**
On line 44, you start `processReviewAsync` but don't await it, then immediately send a response. This creates a "fire and forget" pattern, but it's implemented incorrectly. The catch block on line 46 won't catch errors from the async operation. Consider this pattern:

```typescript
// Option 1: Await the operation
await this.processReviewAsync(repository.owner.login, repository.name, number);
res.status(200).json({ message: 'Review completed' });

// Option 2: Proper fire-and-forget
this.processReviewAsync(repository.owner.login, repository.name, number)
  .catch(error => logger.error('Async review failed:', error));
res.status(200).json({ message: 'Review started' });
return; // Important: prevent further execution
```

**Code Organization:**
Your class-based approach is clean and follows good separation of concerns. The logging is thorough, which will help with debugging. Consider extracting the webhook validation and signature verification into a middleware function that can be reused across all webhook endpoints.

**Testing:**
Given the critical nature of webhook handling, this code needs comprehensive tests covering success cases, error scenarios, and security validations. Focus on testing the signature verification, error handling, and async processing patterns.

**Learning Resources:**
- GitHub Webhook Security: https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks
- Express.js Error Handling: https://expressjs.com/en/guide/error-handling.html
- TypeScript Express Patterns: Look at existing controller patterns in the codebase

Great work on the structured logging and clean separation of concerns! Once the security issues are addressed, this will be a solid foundation for webhook processing.

## ⚡ Action Required

Address security vulnerabilities (especially webhook signature verification) and fix async error handling before merge. Consider adding tests for critical paths.


---
*This review was generated by AI. Please use human judgment for final decisions.*

File: check.ts (hunk 2 above)
Comment (kyleryan207) [inline check.ts:45]: The fire-and-forget pattern here is intentional for webhook processing, but consider adding a response to indicate the async processing has started. This helps with debugging and monitoring webhook deliveries.

File: check.ts (hunk 5 above)
Comment (kyleryan207) [inline check.ts:1]: For webhook controllers, prioritize integration tests that verify the actual webhook payload processing and HTTP response handling. Unit tests for the individual methods are also valuable, especially for error scenarios and edge cases in payload validation.

File: check.ts (hunk 6 above)
Comment (kyleryan207) [inline check.ts:32]: Given this appears to be webhook infrastructure, I'd recommend implementing the repository validation sooner rather than later for security reasons. Webhooks without proper validation can be a security risk if they process requests from unintended repositories.

File: check.ts (hunk 7 above)
Comment (kyleryan207) [inline check.ts:14]: Consider creating a dedicated `types/webhooks.ts` file with properly typed interfaces. This would also allow you to validate the webhook payloads at runtime using libraries like `zod` for additional security.

File: check.ts (hunk 8 above)
Comment (kyleryan207) [inline check.ts:51]: A Result/Either pattern would be ideal here. You could return `{ success: boolean, error?: string, data?: any }` to enable better monitoring and potentially retry logic in the future.

File: check.ts [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        error: error.message,
+        stack: error.stack,
+        owner,
+        repo,
+        pullNumber,
+      });
+    }
+  }
+
+  async handlePullRequestReview(req: Request<{}, {}, PullRequestReviewWebhookEvent>, res: Response): Promise<void> {
```
Comment (kyleryan207) [inline check.ts:75]: Missing webhook signature verification. GitHub webhooks should be validated using the `X-Hub-Signature-256` header to ensure the request is actually from GitHub. This is a critical security requirement for webhook endpoints.

```suggestion
// Add signature validation in constructor or as middleware
private validateWebhookSignature(payload: string, signature: string): boolean {
  const expectedSignature = crypto
    .createHmac('sha256', config.GITHUB_WEBHOOK_SECRET)
    .update(payload)
    .digest('hex');
  return crypto.timingSafeEqual(
    Buffer.from(`sha256=${expectedSignature}`),
    Buffer.from(signature)
  );
}
```

File: check.ts [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      const { action, review, pull_request, repository } = req.body;
+
+      logger.info('Received PR review webhook', {
+        action,
+        reviewState: review?.state,
+        prNumber: pull_request?.number,
+        repo: repository?.full_name,
+      });
+
+      // For now, we'll just log review events
```
Comment (kyleryan207) [inline check.ts:86]: The comment suggests Linear integration, but there's no Linear service imported or configured. Consider adding a placeholder service or removing the comment to avoid confusion about current capabilities.

File: check.ts [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        return;
+      }
+
+      logger.info('Received issue comment webhook', {
+        action,
+        commentId: comment?.id,
+        issueNumber: issue?.number,
+        repo: repository?.full_name,
+      });
+
```
Comment (kyleryan207) [inline check.ts:112]: The comment about command handling suggests this might be a planned feature. Consider adding rate limiting for comment processing to prevent abuse if commands are implemented later.

**Review by kyleryan207 — COMMENTED** [review]
## 🤖 AI Code Review

**Overall Assessment**: ⚠️ NEEDS_CHANGES

**Summary**: This PR implements a webhook controller for GitHub PR reviews, but has critical async handling issues, missing error recovery patterns, and several security concerns that need addressing before it can be safely deployed.

## 📋 Code Quality (5/10)

**Issues**:
- Fire-and-forget async operation in handlePullRequest - processReviewAsync is not awaited
- Missing return statement after res.status(200) calls, could cause 'Cannot set headers after they are sent' errors
- Catch block won't handle async errors from processReviewAsync since it's not awaited
- Missing test coverage for webhook controller - critical for security and reliability
- Commented-out repository validation logic should be implemented or removed
- Missing newline at end of CODEOWNERS file
- Using process.env.npm_package_version may not work in production builds

**Suggestions**:
- Extract webhook request type definitions to improve type safety
- Add proper error handling propagation to processReviewAsync method
- Add logging for non-PR comments for monitoring purposes
- Export controller instance for use in routes
- Follow early return pattern consistently throughout methods

## 📝 Acceptance Criteria
**Status**: ❌ Not Met

While the PR implements basic webhook handling structure, critical security measures and proper async handling are missing. The fire-and-forget pattern for processReviewAsync violates best practices and could lead to silent failures.

## 💬 Detailed Feedback

Welcome to the team! This is a solid start on implementing webhook functionality, but there are some important patterns and security considerations we need to address.

**Critical Issues to Fix:**

1. **Async Handling Pattern**: In `handlePullRequest`, you're calling `processReviewAsync` without awaiting it (line 44). This creates a "fire-and-forget" pattern that can cause several problems:
   - The response is sent immediately without waiting for the review to complete
   - Errors in the async operation won't be caught by your try-catch block
   - You might send the response before the operation fails
   
   **Fix**: Either await the call or implement proper background job processing.

2. **Missing Return Statements**: After sending responses (like line 28), you need to return to prevent further execution. Otherwise, you might try to send multiple responses.

3. **Security Concerns**: This is the most critical area needing attention:
   - **Webhook Verification**: GitHub webhooks should be verified using HMAC-SHA256 signatures to ensure they're actually from GitHub
   - **Repository Validation**: The commented-out validation (lines 32-42) should be implemented to prevent processing unauthorized repositories
   - **Rate Limiting**: Consider adding rate limiting to prevent abuse

**Learning Opportunities:**

Looking at our codebase standards, here are some patterns to follow:
- **Error Handling**: Follow our principle of not making DB calls directly in controllers - wrap them in service functions with clear typing
- **Async Operations**: Based on our async best practices, avoid Promise.all for bulk operations and use sequential processing when appropriate
- **Testing**: Add comprehensive test coverage, especially for security-critical webhook handling

**Helpful Resources:**
- GitHub Webhook Security: https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks
- Express Error Handling: https://expressjs.com/en/guide/error-handling.html

**Specific Code Suggestions:**
```typescript
// Fix the async handling:
async handlePullRequest(req: Request, res: Response): Promise<void> {
  try {
    // ... validation logic
    
    // Option 1: Await the operation
    await this.processReviewAsync(repository.owner.login, repository.name, number);
    return res.status(200).json({ message: 'Review completed' });
    
    // Option 2: Background processing
    // this.processReviewAsync(...).catch(error => logger.error('Background review failed', error));
    // return res.status(202).json({ message: 'Review queued' });
  } catch (error) {
    // Now this will actually catch errors
  }
}
```

You're on the right track with the overall structure! Once these security and async handling issues are addressed, this will be a solid foundation for the webhook system.

## ⚡ Action Required

Please address the security vulnerabilities (especially webhook signature verification), fix the async handling pattern, and add proper error handling before this can be merged. Consider adding test coverage for the critical webhook functionality.


---
*This review was generated by AI. Please use human judgment for final decisions.*
